### DISCUSSION TITLE
String interpolation

### DISCUSSION THREAD

**Discussion by springcomp** [jmespath-community/jmespath.spec#90] [Ideas]
String interpolation is a powerful feature in many modern languages.
It may be worth considering it - in addition of or completely separated from the `format()` function.

I can see value in supporting syntax such as:

`` search( «{greetings}, {dude}!» , { "greetings": "Hello", "dude": "World" } ) -> "Hello, world!" ``

This would need help from the lexer and parser though so might prove overly complex.
The preceding example uses `interpolated-string` because we probably need yet another quotation mark character. Maybe:

```abnf
expression =/ interpolated-string

interpolated-string = "«"  *( interpolated-string-char / interpolating-span  ) "»"
interpolated-string-char = ( %x20-26 / %x28-5B / %x5D-10FFFF ) / interpolated-escape
interpolated-span = "{" expression "}"
interpolated-escape = escape (  "«" / "{" / "}" / "»" )
```

**Comment by innovate-invent** [comment]
I am not sure I like the idea of using untypable syntax. It would be better to simply support quote prefixes like `f""` or `r""`.
What would this provide over the format function other than being more concise?

Where do we draw the line between using syntax for data structure manipulation and functions for data transforms? The arithmetic operators begin to violate that principle, but the argument is that they are such common and useful transforms that it is intuitive to have them as syntax.

**Reply by springcomp** [reply]
Nothing is _untypable_ if you have the [right keyboard](https://github.com/springcomp/optimized-azerty-win) 😁.

Just joking. I think you are right, although I believe modern languages should take advantage of progress and not necessarily restrict themselves to ASCII. That’s why I proposed alternate mathematical symbols in `arithmetic-expressions`.

On another note, I do not necessarily think there is a line to be drawn at all. I think it’s legitimate for languages to evolve and take advantage of new paradigm. First there was `printf()`, then we now have more powerful and type-safe string interpolation in compiled languages.

For the record, can you confirm whether the motivating example referred to in my post above would be possible with the `format()` function using syntax such as:

`` search( format(@, '{greetings}, {dude}!')  , { "greetings": "Hello", "dude": "World" } ) -> "Hello, world!") ``
`` search( format(@, '{}, {}!')  , [ "Hello", "World" ] } ) -> "Hello, world!") ``

**Reply by innovate-invent** [reply]
I do like the idea of richer alternate symbols, I encountered another syntax that makes good use of them while providing reasonable ASCII alternatives. It makes for some really pretty code.

>  I do not necessarily think there is a line to be drawn at all

There is definitely a line, it came up in a few different proposed JEPs and suggestions in the old repos. I'll try to dig up some examples tomorrow.

>  can you confirm whether the motivating example referred to in my post above would be possible with the format() function

Yes, that is the intended functionality.
